Manual, MPR#5780:  nomenclature for existential type names

Octachron opened the pull request:
This pull requests adds a short paragraph in the GADT section of the manual to explain the nomenclature used to name existential type generated when typing GADT.

The nomenclature list is directly inspired from Jacques Garrigue's [comment](http://caml.inria.fr/mantis/view.php?id=5780#c14425) in [MPR#5780](http://caml.inria.fr/mantis/view.php?id=5780) with some added examples to explicit the different items. The last example is quite artificial, and I am open to any more natural and self-contained example.

Octachron committed:
Manual: compiler existential name nomenclature

This commit documents the nomenclature used for naming existential
types when these types are exposed through error messages.

gasche commented:
This `\textquotesingle` business looks painful. Could you rebind it away as `\tq` for example? (I'm not sure which fragment of LaTeX is allowed by I would suppose that `\newcommand` is perfectly fine.)

gasche commented on manual/manual/refman/exten.etex:
```diff
@@ -1315,6 +1315,53 @@ to implement dynamic types.
+\paragraph{Type variable names in error messages}
+The typing of pattern matching in presence of GADT can generate many
+existential types. When reporting a typing error, some of these
+existential types can be exposed and therefore need a
```
I'm not fond of "can be exposed and therefore need a compiler-generated name", as "exposed" does not mean much.  => Maybe "may need to be referred to" instead of "can be exposed", or even "may need to be named in the error explanation" for the whole thing?

Octachron replied:
I have rewritten this sentence to focus more on the error message part.

gasche commented on manual/manual/refman/exten.etex:
```diff
@@ -1315,6 +1315,53 @@ to implement dynamic types.
+The typing of pattern matching in presence of GADT can generate many
+existential types. When reporting a typing error, some of these
+existential types can be exposed and therefore need a
+compiler-generated name. The compiler names these variables accordingly
```
according to

Octachron replied:
Fixed. Thanks.

gasche commented on manual/manual/refman/exten.etex:
```diff
@@ -1315,6 +1315,53 @@ to implement dynamic types.
+\:        The type constructor $Any would escape its scope
+\endcaml
+\item "$'a" is an existential generated by reification of type
+variable "'a" during unification:
```
I found this explanation confusing. I would rather say something like: "an existential variable may also be named `$'foo` if it was unified with the type variable `'foo`".

Octachron replied:
Thanks for the suggestion. I have adopted a compromise to keep the leading "$'a".

gasche commented:
I think it would be nice to say explicitly that this is only the current behavior, and that it may change in the future.

Octachron commented:
The "hand-written" `\caml ... \endcaml` environments were quite painful in general. My prefered solution would be to enable the `caml_example` pseudo-environment,  which is used in the tutorial part of the manual, for `refman/exten.etex`. There are few other places in `refman/exten` that would also benefit for this pseudo-environment. 

I would add a mention that the current nomenclature is incomplete and might be improved in the future.

gasche commented:
On the other hand, using `caml_example` means that if the error-reporting implementation changes, the text will be silently updated on manual rebuild, which can be pretty confusing in a section that is precisely describing the error messages at a given point in time. (Forgetting to update the manual if the errors change is bad, but at least the current text will remain self-consistent.) For the reader of your documentation in source format, `\caml..\endcaml` is not an issue -- `\textquotesingle` on the other hand is a readability issue.

Octachron committed:
Manual: fix existential names

Octachron committed:
Manual: type variable macro for refman/exten

Octachron commented:
I disagree, self-consistent errors are far worse than inconsistencies in my opinion: they are harder to spot, have a longer life expectancy and may even percolate

Anyway, I have added a macro abbreviation `tq` and replaced the previous `\textquotesingle`
by this new abbreviation. 

I have also updated the title of the paragraph to better reflect the paragraph content.

gasche commented:
Thanks!
